lib/pure/os.nim

```diff
@@ -666,6 +589,90 @@ proc isAbsolute*(path: string): bool {.rtl, noSideEffect, extern: "nos$1".} =
   elif defined(posix):
     result = path[0] == '/'
 
+proc parentDir*(path: string): string {.
+  noSideEffect, rtl, extern: "nos$1".} =
+  ## Returns the parent directory of `path`.
+  ##
+  ## Conventions below derive from analogy with the shell:
+  ##   * trailing ``.`` and ``/`` are resolved before taking the parent dir
+  ## It returns empty when attempting to take parent of ``.`` or ``..``
+  ## to indicate it is invalid (and different from ``.``)
+  ##
+  ## The remainder can be obtained with `lastPathPart(path) proc
+  ## <#lastPathPart,string>`_.
+  ##
+  ## See also:
+  ## * `relativePath proc <#relativePath,string,string>`_
+  ## * `splitPath proc <#splitPath,string>`_
+  ## * `tailDir proc <#tailDir,string>`_
+  ## * `parentDirs iterator <#parentDirs.i,string>`_
+  runnableExamples:
+    assert parentDir("") == ""
+    when defined(posix):
+      assert parentDir("/usr/local/bin") == "/usr/local"
+      assert parentDir("foo//bar/") == "foo"
+      assert parentDir("foo") == "."
+      assert parentDir("/foo") == "/"
+      assert parentDir(".") == ""
+  const parentDirOfRootIsEmpty = true # still controversial
```

Comment:
Remove this, it may be controversial but we'll never have ``parentDirOfRootIsEmpty = false``